debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/events/BufferingChangeStreamCursor.java

```diff
@@ -0,0 +1,371 @@
+                noMessageIterations = 0;
+                fetchEvents(cursor);
+            }
+            catch (InterruptedException e) {
+                throw new DebeziumException("Fetcher thread interrupted", e);
+            }
+            finally {
+                cursorRef.set(null);
+                close();
+            }
+        }
+
+        private void fetchEvents(MongoChangeStreamCursor<ChangeStreamDocument<TResult>> cursor) throws InterruptedException {
+            ResumableChangeStreamEvent<TResult> lastEvent = null;
+            var repeat = false;
+            while (isRunning()) {
+                if (!repeat) {
+                    var maybeEvent = fetchEvent(cursor);
+                    if (maybeEvent.isEmpty()) {
+                        LOGGER.warn("Resume token not available on this poll");
+                        continue;
+                    }
+                    lastEvent = maybeEvent.get();
+                }
+                repeat = !enqueue(lastEvent);
+            }
+        }
+
+        private Optional<ResumableChangeStreamEvent<TResult>> fetchEvent(MongoChangeStreamCursor<ChangeStreamDocument<TResult>> cursor) {
+            var beforeEventPollTime = clock.currentTimeAsInstant();
+            var document = cursor.tryNext();
+            metrics.onSourceEventPolled(document, clock, beforeEventPollTime);
+            throttleIfNeeded(document);
+
+            // Only create resumable event if we have either document or cursor resume token
+            // Cursor resume token may be `null` in case of issues like SERVER-63772, and situations called out in the Javadocs:
+            // > resume token [...] can be null if the cursor has either not been iterated yet, or the cursor is closed.
+            return Optional.<ResumableChangeStreamEvent<TResult>> empty()
+                    .or(() -> Optional.ofNullable(document).map(ResumableChangeStreamEvent::new))
+                    .or(() -> Optional.ofNullable(cursor.getResumeToken()).map(ResumableChangeStreamEvent::new));
+        }
+
+        private void throttleIfNeeded(ChangeStreamDocument<TResult> document) {
+            if (document == null) {
+                noMessageIterations++;
+            }
+            if (noMessageIterations >= THROTTLE_NO_MESSAGE_BEFORE_PAUSE) {
+                LOGGER.debug("Sleeping after {} empty polls", noMessageIterations);
+                throttler.sleepWhen(true);
+                noMessageIterations = 0;
+            }
+        }
+
+        private boolean enqueue(ResumableChangeStreamEvent<TResult> event) throws InterruptedException {
+            var available = this.capacity.tryAcquire(QUEUE_OFFER_TIMEOUT_MS, TimeUnit.MILLISECONDS);
+            if (!available) {
+                LOGGER.warn("Unable to acquire buffer lock, buffer queue is likely full");
+                return false;
+            }
+            // always true
+            return queue.offer(event);
+        }
+    }
```

Comment:
@jpechane I think static value for the timeout should be OK, no? We can always make it configurable in the future